hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java

```diff
@@ -2878,27 +2880,41 @@ private void copyFile(String srcKey, String dstKey, long size)
       }
     };
 
-    once("copyFile(" + srcKey + ", " + dstKey + ")", srcKey,
-        () -> {
-          ObjectMetadata srcom = getObjectMetadata(srcKey);
-          ObjectMetadata dstom = cloneObjectMetadata(srcom);
-          setOptionalObjectMetadata(dstom);
-          CopyObjectRequest copyObjectRequest =
-              new CopyObjectRequest(bucket, srcKey, bucket, dstKey);
-          setOptionalCopyObjectRequestParameters(copyObjectRequest);
-          copyObjectRequest.setCannedAccessControlList(cannedACL);
-          copyObjectRequest.setNewObjectMetadata(dstom);
-          Copy copy = transfers.copy(copyObjectRequest);
-          copy.addProgressListener(progressListener);
-          try {
-            copy.waitForCopyResult();
-            incrementWriteOperations();
-            instrumentation.filesCopied(1, size);
-          } catch (InterruptedException e) {
-            throw new InterruptedIOException("Interrupted copying " + srcKey
-                + " to " + dstKey + ", cancelling");
-          }
-        });
+    try {
+      return once("copyFile(" + srcKey + ", " + dstKey + ")", srcKey,
+          () -> {
+            ObjectMetadata srcom = getObjectMetadata(srcKey);
+            ObjectMetadata dstom = cloneObjectMetadata(srcom);
+            setOptionalObjectMetadata(dstom);
+            CopyObjectRequest copyObjectRequest =
+                new CopyObjectRequest(bucket, srcKey, bucket, dstKey);
+            setOptionalCopyObjectRequestParameters(copyObjectRequest);
+            copyObjectRequest.setCannedAccessControlList(cannedACL);
+            copyObjectRequest.setNewObjectMetadata(dstom);
+            String id = srcom.getVersionId();
+            if (id != null) {
+              copyObjectRequest.setSourceVersionId(id);
+            } else if (isNotEmpty(srcom.getETag())) {
+              copyObjectRequest.withMatchingETagConstraint(srcom.getETag());
+            }
+            Copy copy = transfers.copy(copyObjectRequest);
+            copy.addProgressListener(progressListener);
+            try {
+              CopyResult r = copy.waitForCopyResult();
+              incrementWriteOperations();
+              instrumentation.filesCopied(1, size);
+              return r;
+            } catch (InterruptedException e) {
+              throw (IOException) new InterruptedIOException(
```

Comment:
Why do you cast `InterruptedIOException` to `IOException`? If you cast it, then this method won't throw InterruptedIOException, just `IOException`, so it can be removed from it's signature.